devel/validate-platforms

```diff
@@ -0,0 +1,101 @@
+#!/bin/bash
+#
+# Validate different platform builds of the final Nextstrain image.
+#
+set -euo pipefail
+
+# Set default values.
+registry=localhost:5000
+tag=latest
+
+# Read command-line arguments.
+while getopts "r:t:" opt; do
+    case "$opt" in
+        r) registry="$OPTARG";;
+        t) tag="$OPTARG";;
+        *) echo "Usage: $0 [-r <registry>] [-t <tag>]" 1>&2; exit 1;;
+    esac
+done
+
+IMAGE="$registry/nextstrain/base:$tag"
+PLATFORMS=(linux/amd64 linux/arm64)
+
+main() {
+    # Check software where downloaded versions are not deterministic at
+    # different points in time (e.g. if "latest" is downloaded).
+
+    local report_dir
+    report_dir="$(mktemp -dt "$(basename "$0")")"
```

Comment:
_blocking_

```suggestion
    report_dir="$(mktemp -dt "$(basename "$0")"-XXXXXX)"
```

The X's in my patch were intentional; they're a required part of the template for GNU coreutils' mktemp:

```
$ ./devel/validate-platforms -r docker.io
mktemp: too few X's in template ‘validate-platforms’
```

from mktemp(1):

> TEMPLATE must contain at least 3 consecutive 'X's in last component.

Reply:
I noticed when CI failed... looks like another [difference between macOS and other OS's](https://unix.stackexchange.com/q/30091/181911) 😕

On a GitHub Codespace instance:

```
$ uname -a
Linux codespaces-cf4b1b 5.4.0-1103-azure #109~18.04.1-Ubuntu SMP Wed Jan 25 20:53:00 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

$ mktemp --version
mktemp (GNU coreutils) 8.30
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Jim Meyering and Eric Blake.

$ man mktemp | tail -n 1
GNU coreutils 8.30                   September 2019                            MKTEMP(1)

$ mktemp -dt "validate-platforms-XXX"
/tmp/validate-platforms-Ege
```

On macOS Ventura:

```
% mktemp --version
mktemp: illegal option -- -
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 

% man mktemp | tail -n 1
macOS 13.1                      August 11, 2014                     macOS 13.1

% mktemp -dt "validate-platforms-XXX"
/var/folders/m5/30jq1dcn439_lzq16twm_jcc0000gn/T/validate-platforms-XXX.bdPORZe5
```

The macOS version doesn't replace the X's but it still creates a valid temp directory, so I'll add them back in.

Curious, any reason you chose 6 X's over the minimum 3?

Reply:
> The macOS version doesn't replace the X's but it still creates a valid temp directory, so I'll add them back in.

Yep. It doesn't replace the Xs in `-t prefix` because it treats that as a "prefix" (to which it prepends the temp dir and appends the random chars) as opposed to a "template", e.g. the difference between these two usages:

```
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
```

> Curious, any reason you chose 6 X's over the minimum 3?

Habit because it's more secure, though that's hardly a concern in this usage.

Reply:
X's added in [force-push](https://github.com/nextstrain/docker-base/compare/468cb918971f161fce0f9df54c34770614a2f9ee..c70aa6c2d37aed5572ae20f1fdfbd0c287559bb7).